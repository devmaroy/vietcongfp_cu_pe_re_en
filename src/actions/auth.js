// Auth Actions

import { firebase } from '../firebase/firebase';
import history from '../routers/history';
import { 
    LOGOUT, 
    SET_CURRENT_USER, 
    SET_AUTH_LOADING,
    CLEAR_AUTH_LOADING,
    GET_ERRORS
} from './types';

// Set current user helper function
const setUserData = ( user, dispatch ) => {
    const userData = {
        uid: user.uid,
        name: user.displayName,
        email: user.email
    };

    dispatch( setCurrentUser( userData ) );
};


// Login
export const startLogin = ( { email, password } ) => {
    return ( dispatch ) => {
        dispatch( setAuthLoading() );

        return firebase
            .auth()
            .signInWithEmailAndPassword( email, password )
            .then( ( user ) => {  
                setUserData( user, dispatch );
                history.push( '/' );
            })
            .catch( ( err ) => {
                dispatch({
                    type: GET_ERRORS,
                    payload: err 
                });

                dispatch( clearAuthLoading() );
            });
    };
};


// Register
export const startRegister = ( { name, email, password } ) => {
    return ( dispatch ) => {
        dispatch( setAuthLoading() );
        
            return firebase
            .auth()
            .createUserWithEmailAndPassword( email, password )
            .then( ( user ) => {
                if ( user ) {
                    user.updateProfile({
                        displayName: name
                    }).then(() => {
                        setUserData( user, dispatch );
                        history.push( '/' );
                    });
                }
            })
            .catch( ( err ) => {
                dispatch({
                    type: GET_ERRORS,
                    payload: err
                });

                dispatch( clearAuthLoading() );
            });
    };
};


// Logout
export const startLogout = () => {
    return ( dispatch ) => {
        
        return firebase
            .auth()
            .signOut()
            .then(() => {
                dispatch( logout() );
                history.push( '/login' );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Logout
export const logout = () => ({
    type: LOGOUT
});


// Set Current User
export const setCurrentUser = ( user ) => ({
    type: SET_CURRENT_USER,
    user
});


// Set Auth Loading
export const setAuthLoading = () => ({
    type: SET_AUTH_LOADING
});


// Clear Auth Loading
export const clearAuthLoading = () => ({
    type: CLEAR_AUTH_LOADING
});