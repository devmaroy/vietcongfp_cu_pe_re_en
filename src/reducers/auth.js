// Auth Reducer

import { 
    SET_CURRENT_USER, 
    SET_AUTH_LOADING, 
    CLEAR_AUTH_LOADING,
    LOGOUT 
} from '../actions/types';

const authReducerDefaultState = {
    isLoading: false,
    isAuthenticated: false,
    user: {}
};

export default ( state = authReducerDefaultState, action ) => {
    switch ( action.type ) {
        case SET_CURRENT_USER:
            return {
                ...state,
                isLoading: false,
                isAuthenticated: !!action.user.uid,
                user: action.user
            }

        case SET_AUTH_LOADING:
            return {
                ...state,
                isLoading: true
            }

        case CLEAR_AUTH_LOADING:
            return {
                ...state,
                isLoading: false
            }

        case LOGOUT:
            return {};
        default:
            return state;
    }
};