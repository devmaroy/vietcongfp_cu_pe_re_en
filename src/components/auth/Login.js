import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import validator from 'validator';
import { Link } from 'react-router-dom';
import { startLogin } from '../../actions/auth';
import TextFieldGroup from '../common/TextFieldGroup';
import Spinner from '../common/Spinner';

export class Login extends React.Component {
    state = {
        email: '',
        password: '',
        errors: {}
    };

    componentWillReceiveProps( nextProps ) {
        if ( nextProps.errors ) {
            const errors = {};
            errors[ 'authError' ] = nextProps.errors.message;

            this.setState( { errors } );
        }
    };

    handleValidation = () => {
        const errors = {};
        let formIsValid = true;

        // Email
        if ( ! this.state.email ) {
            formIsValid = false;
            errors[ 'email' ] = 'Email cannot be empty.';
        } else if ( ! validator.isEmail( this.state.email ) ) {
            formIsValid = false;
            errors[ 'email' ] = 'Email is invalid.';
        }

        // Password
        if ( ! this.state.password ) {
            formIsValid = false;
            errors[ 'password' ] = 'Password cannot be empty.';
        }

        this.setState( () => ( { errors } ) );

        return formIsValid;
    };

    handleChange = ( e ) => {
        const target = e.target;
        const value = target.value;
        const name = target.name;
    
        this.setState( () => ( { [ name ]: value } ) );
    };

    onSubmit = ( e ) => {
        e.preventDefault();

        if ( this.handleValidation() ) {
            const { email, password } = this.state;
            const loginData = { email, password };

            this.props.startLogin( loginData );
        }
    };

    render() {
        const { errors } = this.state;
        const { isLoading } = this.props.auth;

        let loginContent;

        if ( isLoading ) {
            loginContent = <Spinner />;
        } else {
            loginContent = (
                <div>
                    <h2>Sign In</h2>
                    <form onSubmit={ this.onSubmit } className="form">
                        { errors.authError && <p>{ errors.authError }</p> }
                        <div className="input-group">
                            <div className="input-group__item">
                                <TextFieldGroup 
                                    name="email"
                                    placeholder="Your email"
                                    value={ this.state.email }
                                    onChange={ this.handleChange }
                                    error={ errors.email }
                                    className="text-input"
                                />
                            </div>
                        </div>
                        <div className="input-group">
                            <div className="input-group__item">
                                <TextFieldGroup 
                                    type="password"
                                    name="password"
                                    placeholder="Your password"
                                    value={ this.state.password }
                                    onChange={ this.handleChange }
                                    error={ errors.password }
                                    className="text-input"
                                />
                            </div>
                        </div>
                        <div className="form-meta">
                            <p>Don't have an account? <Link to="/register">Sign Up</Link></p>
                            <button className="button">Sign In</button>
                        </div>
                    </form>
                </div>
            );
        }

        return (
            <div className="content-container">
                <div className="page">
                    { loginContent }
                </div>
            </div>
        );
    };
};

Login.propTypes = {
    auth: PropTypes.object.isRequired,
    errors: PropTypes.object,
    startLogin: PropTypes.func.isRequired
};

const mapStateToProps = ( state ) => ({
    auth: state.auth,
    errors: state.errors
});

const mapDispatchToProps = ( dispatch ) => ({
    startLogin: ( loginData ) => dispatch( startLogin( loginData ) ) 
});

export default connect( mapStateToProps, mapDispatchToProps )( Login );