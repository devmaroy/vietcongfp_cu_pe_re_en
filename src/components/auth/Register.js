import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import validator from 'validator';
import { Link } from 'react-router-dom';
import { startRegister } from '../../actions/auth';
import TextFieldGroup from '../common/TextFieldGroup';
import Spinner from '../common/Spinner';

export class Register extends React.Component {
    state = {
        name: '',
        email: '',
        password: '',
        passwordConfirm: '',
        errors: {}
    };

    componentWillReceiveProps( nextProps ) {
        if ( nextProps.errors ) {
            const errors = {};
            errors[ 'authError' ] = nextProps.errors.message;

            this.setState( { errors } );
        }
    };

    handleValidation = () => {
        const errors = {};
        let formIsValid = true;

        // Name
        if ( ! this.state.name ) {
            formIsValid = false;
            errors[ 'name' ] = 'Name cannot be empty.';
        }

        // Email
        if ( ! this.state.email ) {
            formIsValid = false;
            errors[ 'email' ] = 'Email cannot be empty.';
        } else if ( ! validator.isEmail( this.state.email ) ) {
            formIsValid = false;
            errors[ 'email' ] = 'The email address is badly formatted.';
        }

        // Password
        if ( ! this.state.password ) {
            formIsValid = false;
            errors[ 'password' ] = 'Password cannot be empty.';
        } else if ( this.state.password.length < 6 ) {
            formIsValid = false;
            errors[ 'password' ] = 'The password must be 6 characters long or more.';
        }

        // Password confirm
        if ( ! this.state.passwordConfirm ) {
            formIsValid = false;
            errors[ 'passwordConfirm' ] = 'Password confirm cannot be empty.';
        } else if ( this.state.passwordConfirm !== this.state.password ) {
            formIsValid = false;
            errors[ 'passwordConfirm' ] = 'Password confirm do not match with your password.';
        }

        this.setState( () => ( { errors } ) );

        return formIsValid;
    };

    handleChange = ( e ) => {
        const target = e.target;
        const value = target.value;
        const name = target.name;
    
        this.setState( () => ( { [ name ]: value } ) );
    };

    onSubmit = ( e ) => {
        e.preventDefault();

        if ( this.handleValidation() ) {
            const { name, email, password, passwordConfirm } = this.state;
            const registerData = { name, email, password, passwordConfirm };

            this.props.startRegister( registerData );
        }
    };

    render() {
        const { errors } = this.state;
        const { isLoading } = this.props.auth;

        let registerContent;

        if ( isLoading ) {
            registerContent = <Spinner />;
        } else {
            registerContent = (
                <div>
                    <h2>Sign Up</h2>
                    <form onSubmit={ this.onSubmit } className="form">
                        { errors.authError && <p>{ errors.authError }</p> }
                        <div className="input-group">
                            <div className="input-group__item">
                                <TextFieldGroup 
                                    name="name"
                                    placeholder="Your name"
                                    value={ this.state.name }
                                    onChange={ this.handleChange }
                                    error={ errors.name }
                                    className="text-input"
                                />
                            </div>
                        </div>
                        <div className="input-group">
                            <div className="input-group__item">                                
                                <TextFieldGroup 
                                    type="email"
                                    name="email"
                                    placeholder="Your email"
                                    value={ this.state.email }
                                    onChange={ this.handleChange }
                                    error={ errors.email }
                                    className="text-input"
                                />
                            </div>
                        </div>
                        <div className="input-group">
                            <div className="input-group__item">
                                <TextFieldGroup 
                                    type="password"
                                    name="password"
                                    placeholder="Your password"
                                    value={ this.state.password }
                                    onChange={ this.handleChange }
                                    error={ errors.email }
                                    className="text-input"
                                />
                            </div>
                        </div>
                        <div className="input-group">
                            <div className="input-group__item">
                                <TextFieldGroup 
                                    type="password"
                                    name="passwordConfirm"
                                    placeholder="Confirm your password"
                                    value={ this.state.passwordConfirm }
                                    onChange={ this.handleChange }
                                    error={ errors.email }
                                    className="text-input"
                                />
                            </div>
                        </div>
                        <div className="form-meta">
                            <p>Already have an account? <Link to="/login">Log In</Link></p>
                            <button className="button">Sign Up</button>
                        </div>
                    </form>
                </div>
            );
        }

        return (
            <div className="content-container">
                <div className="page">
                    { registerContent }
                </div>
            </div>
        );
    };
};

Register.propTypes = {
    auth: PropTypes.object.isRequired,
    errors: PropTypes.object,
    startRegister: PropTypes.func.isRequired
};

const mapStateToProps = ( state ) => ({
    auth: state.auth,
    errors: state.errors
});

const mapDispatchToProps = ( dispatch ) => ({
    startRegister: ( registerData ) => dispatch( startRegister( registerData ) ) 
});

export default connect( mapStateToProps, mapDispatchToProps )( Register );