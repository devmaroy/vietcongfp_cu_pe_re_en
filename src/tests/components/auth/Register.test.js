import React from 'react';
import { shallow } from 'enzyme';
import { Register } from '../../../components/auth/Register';

test( 'should render correctly Register with spinner', () => {
    const wrapper = shallow( 
         <Register 
            startRegister={ () => {} }
            auth={ { isLoading: true } }
         />
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find( 'Spinner' ).exists() ).toBeTruthy();
});

test( 'should render correctly Login without spinner', () => {
    const wrapper = shallow( 
        <Register 
            startRegister={ () => {} }
            auth={ { isLoading: false } }
        />
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find( 'Spinner' ).exists() ).toBeFalsy();
});


test( 'should correctly render form inside Login', () => {
    const wrapper = shallow(
        <Register
            startRegister={ () => {} }
            auth={ { isLoading: false } }
        />  
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find( 'form' ).exists() ).toBeTruthy();
});

test( 'should correctly render error object if no data provided', () => {
    const wrapper = shallow(
        <Register 
            startRegister={ () => {} }
            auth={ { isLoading: false } }
        />
    );

    wrapper.find( 'form' ).simulate( 'submit', {
        preventDefault: () => {}
    });

    expect( wrapper.state( 'errors' ).name.length ).toBeGreaterThan( 0 );
    expect( wrapper.state( 'errors' ).email.length ).toBeGreaterThan( 0 );
    expect( wrapper.state( 'errors' ).password.length ).toBeGreaterThan( 0 );
    expect( wrapper.state( 'errors' ).passwordConfirm.length ).toBeGreaterThan( 0 );
});

test( 'should set name on input change', () => {
    const value = 'John';

    const wrapper = shallow( 
        <Register 
            startRegister={ () => {} }
            auth={ { isLoading: false } }
        />
    );

    const event = { target: { name: 'name', value } };

    wrapper.find( 'TextFieldGroup' ).at( 0 ).simulate( 'change', event );

    expect( wrapper.state( 'name' ) ).toBe( value );
});

test( 'should set email on input change', () => {
    const value = 'emailo@gmailo.com';

    const wrapper = shallow( 
        <Register 
            startRegister={ () => {} }
            auth={ { isLoading: false } }
        />
    );

    const event = { target: { name: 'email', value } };

    wrapper.find( 'TextFieldGroup' ).at( 1 ).simulate( 'change', event );

    expect( wrapper.state( 'email' ) ).toBe( value );
});

test( 'should set password on input change', () => {
    const value = 'password15';

    const wrapper = shallow( 
        <Register 
            startRegister={ () => {} }
            auth={ { isLoading: false } }
        />
    );

    const event = { target: { name: 'password', value } };

    wrapper.find( 'TextFieldGroup' ).at( 2 ).simulate( 'change', event );

    expect( wrapper.state( 'password' ) ).toBe( value );
});

test( 'should set password confirm on input change', () => {
    const value = 'password15';

    const wrapper = shallow( 
        <Register 
            startRegister={ () => {} }
            auth={ { isLoading: false } }
        />
    );

    const event = { target: { name: 'passwordConfirm', value } };

    wrapper.find( 'TextFieldGroup' ).at( 3 ).simulate( 'change', event );

    expect( wrapper.state( 'passwordConfirm' ) ).toBe( value );
});

test( 'should call startRegister', () => {
    const registerData = { 
        name: 'John', 
        email: 'correct@email.com', 
        password: 'password123456', 
        passwordConfirm: 'password123456' 
    };

    const startRegister = jest.fn();

    const wrapper = shallow(
        <Register 
            startRegister={ startRegister }
            auth={ { isLoading: false } }
        />
    );

    wrapper.setState( {
        name: registerData.name,
        email: registerData.email, 
        password: registerData.password,
        passwordConfirm: registerData.passwordConfirm
    });

    wrapper.find( 'form' ).simulate( 'submit', {
        preventDefault: () => {}
    });    

    expect( startRegister ).toHaveBeenLastCalledWith( registerData );
});