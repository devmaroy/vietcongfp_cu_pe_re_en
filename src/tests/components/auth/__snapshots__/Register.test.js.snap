// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render form inside Login 1`] = `
<div
  className="content-container"
>
  <div
    className="page"
  >
    <div>
      <h2>
        Sign Up
      </h2>
      <form
        className="form"
        onSubmit={[Function]}
      >
        <div
          className="input-group"
        >
          <div
            className="input-group__item"
          >
            <TextFieldGroup
              className="text-input"
              name="name"
              onChange={[Function]}
              placeholder="Your name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="input-group"
        >
          <div
            className="input-group__item"
          >
            <TextFieldGroup
              className="text-input"
              name="email"
              onChange={[Function]}
              placeholder="Your email"
              type="email"
              value=""
            />
          </div>
        </div>
        <div
          className="input-group"
        >
          <div
            className="input-group__item"
          >
            <TextFieldGroup
              className="text-input"
              name="password"
              onChange={[Function]}
              placeholder="Your password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          className="input-group"
        >
          <div
            className="input-group__item"
          >
            <TextFieldGroup
              className="text-input"
              name="passwordConfirm"
              onChange={[Function]}
              placeholder="Confirm your password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          className="form-meta"
        >
          <p>
            Already have an account? 
            <Link
              replace={false}
              to="/login"
            >
              Log In
            </Link>
          </p>
          <button
            className="button"
          >
            Sign Up
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`should render correctly Login without spinner 1`] = `
<div
  className="content-container"
>
  <div
    className="page"
  >
    <div>
      <h2>
        Sign Up
      </h2>
      <form
        className="form"
        onSubmit={[Function]}
      >
        <div
          className="input-group"
        >
          <div
            className="input-group__item"
          >
            <TextFieldGroup
              className="text-input"
              name="name"
              onChange={[Function]}
              placeholder="Your name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="input-group"
        >
          <div
            className="input-group__item"
          >
            <TextFieldGroup
              className="text-input"
              name="email"
              onChange={[Function]}
              placeholder="Your email"
              type="email"
              value=""
            />
          </div>
        </div>
        <div
          className="input-group"
        >
          <div
            className="input-group__item"
          >
            <TextFieldGroup
              className="text-input"
              name="password"
              onChange={[Function]}
              placeholder="Your password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          className="input-group"
        >
          <div
            className="input-group__item"
          >
            <TextFieldGroup
              className="text-input"
              name="passwordConfirm"
              onChange={[Function]}
              placeholder="Confirm your password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          className="form-meta"
        >
          <p>
            Already have an account? 
            <Link
              replace={false}
              to="/login"
            >
              Log In
            </Link>
          </p>
          <button
            className="button"
          >
            Sign Up
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`should render correctly Register with spinner 1`] = `
<div
  className="content-container"
>
  <div
    className="page"
  >
    <Spinner
      alt="Loading..."
      imagePath="/images/loader.gif"
      style={
        Object {
          "display": "block",
          "margin": "60px",
          "width": "80px",
        }
      }
    />
  </div>
</div>
`;
